### ISSUE TITLE
Add more detail to commands.md documentation

### ISSUE THREAD

**Issue by leftler** [microsoft/azure-pipelines-tasks#1867]
The documentation located at https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md needs more detail added to it, here are some things that I am still unsure of after trying to puzzle it out for a while

All the issues I have found mainly have been with the documentation of `##vso[task.logdetail]`
- Does `parentid` need to be passed with every request or is it a write once field like `name` or `order`?
- What format should time be in for `starttime` and `finishtime`? If it is a format without timezone should it be local time or UTC?
- What format should the GUIDs be in?
- What should we be putting in for `type`? The two examples that you show use `build` as the value passed in, should we always pass in `build` or are there other valid values?
- What format does `progress` need to be in, can it accept decimals, can it accept commas?
- Is there well defined behavior when two objects have the same value for `order`?
- Examples are not actual examples: `##vso[task.logdetail id=new guid;name=project1;type=build;order=1]create new timeline record.` would fail because `id=new guid` is not a valid value.

**Comment by leftler** [comment]
Also, it is not clear where "current operation" is used, or how doing parent id's provides any benefit. Neither is visible. 

For example in the below image G7x Build is the parent of all other items in the list, also the last thing that happened to it should have set the current operation to "Handle SeclectedProjects", however there is 0 indication of any kind of hierarchy and the only place I could see "Handle SeclectedProjects" was in the agent console (and even then it did not include then name, it just wrote ": Handle SelectProjects")

![capture](https://cloud.githubusercontent.com/assets/1425499/16020059/d7f105ba-3171-11e6-8ba7-e5f83c13eb65.PNG)

**Comment by bgever** [comment]
I'm also struggling with the documentation. Can I use the `type` parameter to log Test results? Is `test` a valid record type? I'd like to write a reporter for [Mocha test](https://mochajs.org/) progress.

**Comment by Kujawadl** [comment]
Can we get an update on this? I can't find any other documentation on these commands.

I can't get anything using task.logdetail working...

**Comment by logicbomb421** [comment]
I too would like an update here. There appear to be a lot of useful vso commands we can utilize, but as the OP stated, the documentation is really lacking. The few commands I currently utilize I got working via trial and error.

**Comment by redshiftz** [comment]
What is the status of this?

**Comment by matthid** [comment]
Yes I agree this needs to added to the docs. For now you can peek into the code and answer some of the questions:

https://github.com/Microsoft/vsts-agent/blob/59b25208aa1c17a30c1dddc4828044fd89513599/src/Agent.Worker/TaskCommandExtension.cs#L84-L237


> Does parentid need to be passed with every request or is it a write once field like name or order?

No it can be left out. If it is given it needs to be equal to the first usage.

> What format should time be in for starttime and finishtime? If it is a format without timezone should it be local time or UTC?

`DateTime.TryParse(dateTimeText, CultureInfo.CurrentCulture, DateTimeStyles.AdjustToUniversal, out dateTime)` is used:

https://github.com/Microsoft/vsts-agent/blob/59b25208aa1c17a30c1dddc4828044fd89513599/src/Agent.Worker/TaskCommandExtension.cs#L611-L620

> What format should the GUIDs be in?

`new Guid(...)` is used.

> What should we be putting in for type? The two examples that you show use build as the value passed in, should we always pass in build or are there other valid values?

It looks like a custom string.

> What format does progress need to be in, can it accept decimals, can it accept commas?

`Int32.TryParse(percentCompleteValue, out progress)` is used.

**Comment by d3r3kk** [comment]
I don't see any sort of sub-timeline information anywhere in the latest version of the Build output UX.

In the newest UX for builds, where does the output of logdetail go? Is this even relevant information any longer?

**Comment by MaciejSzczepanski** [comment]
I was hoping it will be improved in new Azure Devops server 2019, but it seems like it was removed. I don't see any timeline view... Command ##vso[task.logdetail is basically useless and should be removed from the documenation...

**Comment by tillig** [comment]
I just ran into this myself, wanting to provide some detail about the state of some long running tasks.

It appears [this issue](https://github.com/microsoft/azure-pipelines-tasks/issues/5961) noted that when the build log display changed, the timeline with this info disappeared. The author of the issue was directed to take it up with the [Developer Community website](https://developercommunity.visualstudio.com/content/idea/post.html?space=21) which isn't great as far as explaining _why_ this doesn't work as expected... or really what to expect.

Effectively the same question has been asked [on the docs repo](https://github.com/MicrosoftDocs/vsts-docs/issues/5263) to see if the docs could add clarification around what the expected outcome of `task.logdetail` usage is... but again, the author was punted to ask on StackOverflow or the Developer Community website.

The only answered StackOverflow question I can find [is here](https://stackoverflow.com/questions/58231828/where-did-the-build-timeline-tab-go-in-azure-devops-server), which basically says this isn't available in the "new build output." There's a nearly identical answer [on this question](https://stackoverflow.com/questions/51611259/can-i-write-custom-task-start-end-records-to-vsts-build-release-log) which wasn't accepted.

In all of these cases, the "answer" seems to be "someone should suggest that get implemented," but I think it's valid to have - somewhere in the documentation - that with the "new build output" view _the timeline is no longer a thing that should be expected._ Whether it should be re-implemented, sure, may be something that should be voted on, but man, it'd be nice to know that it's not just me unable to find where my cheese was moved.

**Comment by tillig** [comment]
As a follow up, apparently this [was suggested on the Developer Community site](https://developercommunity.visualstudio.com/idea/381135/vsotasklogdetail-should-show-start-time-finish-tim.html) and was closed due to 90 days of inactivity. It was [reported a second time](https://developercommunity.visualstudio.com/content/problem/341733/per-task-timeline-in-new-preview-ui.html) and summarily closed with the Microsoft representative pretty clearly not understanding what was being asked.

I'm not sure how we might be expected to make any traction on getting docs updated or getting the view fixed when it doesn't seem anyone's actually listening to the issue.

**Comment by chwarr** [comment]
This issue still exists as more-or-less as originally reported. It should not be marked as stale. The next action needs to be taken by the Azure Pipelines Task team to clarify the format and semantics here. This is not something that we, the users of Azure Pipelines, can definitively answer ourselves.

**Comment by jnm2** [comment]
Hi, the outstanding need has not been addressed. The bot should not close this issue without a response from the team which sets some expectations.

**Comment by snodecoder** [comment]
The following information could be added for example. I could not find this anywhere in the official documentation.
Thanks @ [https://developercommunity.visualstudio.com/t/write-host-foreground-color-with-powershell-task-i/440605](url)

##[error] : Highlight the line in the log red
##[warning] : Highlight the line in the log orange
##[section] : Highlight the line in the log green
##[command] : Highlight the line in the log blue
##[debug] : Highlight the line in the log purple

For example for Powershell you would use the following in your script:
> Write-Host "##[section] This is colored green!"
Write-Host "##[command] This is colored blue!"
Write-Host "##[debug] This is colored purple!"
Write-Host "##[warning] This is colored orange!"
Write-Host "##vso[task.logissue type=warning;]This is colored orange!" 
Write-Host "##[error] This is colored red!"
Write-Host "##vso[task.logissue type=error;]This is colored red!"

**Comment by alexander-smolyakov** [comment]
@snodecoder, information regarding logging commands and text highlighting already provided on the [related page](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands) in the ms docs.

**Comment by snodecoder** [comment]
> @snodecoder, information regarding logging commands and text highlighting already provided on the [related page](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands) in the ms docs.

My bad ;)
Thanks!

**Comment by jnm2** [comment]
Again:

> Hi, the outstanding need has not been addressed. The bot should not close this issue without a response from the team which sets some expectations.

**Comment by jnm2** [comment]
Again:

> Hi, the outstanding need has not been addressed. The bot should not close this issue without a response from the team which sets some expectations.

**Comment by chwarr** [comment]
This issue _is not stale_: we're still waiting on the Azure DevOps Pipeline team to address the gaps in the documentation.

The bot logic needs to be updated to not mark things as stale when the pending action is on the Pipelines teams itself.

**Comment by jnm2** [comment]
Who needs to be pinged in order to get a response?

**Comment by chwarr** [comment]
This issue is not stale: we're _still_ waiting on the Azure DevOps Pipeline team to address the gaps in the documentation.

The bot logic needs to be updated to not mark things as stale when the pending action is on the Pipelines teams itself.

**Comment by KonstantinTyukalov** [comment]
Hi @chwarr! We have documentation about the `task.logdetail` command here:
https://learn.microsoft.com/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logdetail-create-or-update-a-timeline-record-for-a-task

**Comment by chwarr** [comment]
The documentation at https://learn.microsoft.com/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logdetail-create-or-update-a-timeline-record-for-a-task has the same issues as in the original report. This is a request to provide reference-level documentation for everything that logging can do.
